engineio/src/asynchronous/async_socket.rs

```diff
@@ -0,0 +1,219 @@
+use std::{
+    fmt::Debug,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+
+use bytes::Bytes;
+use tokio::{
+    runtime::Handle,
+    sync::{Mutex, RwLock},
+    time::Instant,
+};
+
+use crate::{
+    asynchronous::{callback::OptionalCallback, transport::AsyncTransportType},
+    error::Result,
+    packet::{HandshakePacket, Payload},
+    Error, Packet, PacketId,
+};
+
+#[derive(Clone)]
+pub struct Socket {
+    handle: Handle,
+    transport: Arc<AsyncTransportType>,
+    on_close: OptionalCallback<()>,
+    on_data: OptionalCallback<Bytes>,
+    on_error: OptionalCallback<String>,
+    on_open: OptionalCallback<()>,
+    on_packet: OptionalCallback<Packet>,
+    connected: Arc<AtomicBool>,
+    last_ping: Arc<Mutex<Instant>>,
+    last_pong: Arc<Mutex<Instant>>,
+    connection_data: Arc<HandshakePacket>,
+    /// Since we get packets in payloads it's possible to have a state where only some of the packets have been consumed.
+    remaining_packets: Arc<RwLock<Option<crate::packet::IntoIter>>>,
+}
+
+impl Socket {
+    pub(crate) fn new(
+        transport: AsyncTransportType,
+        handshake: HandshakePacket,
+        on_close: OptionalCallback<()>,
+        on_data: OptionalCallback<Bytes>,
+        on_error: OptionalCallback<String>,
+        on_open: OptionalCallback<()>,
+        on_packet: OptionalCallback<Packet>,
+    ) -> Self {
+        Socket {
+            handle: Handle::current(),
+            on_close,
+            on_data,
+            on_error,
+            on_open,
+            on_packet,
+            transport: Arc::new(transport),
+            connected: Arc::new(AtomicBool::default()),
+            last_ping: Arc::new(Mutex::new(Instant::now())),
+            last_pong: Arc::new(Mutex::new(Instant::now())),
+            connection_data: Arc::new(handshake),
+            remaining_packets: Arc::new(RwLock::new(None)),
+        }
+    }
+
+    /// Opens the connection to a specified server. The first Pong packet is sent
+    /// to the server to trigger the Ping-cycle.
+    pub async fn connect(&self) -> Result<()> {
+        // SAFETY: Has valid handshake due to type
+        self.connected.store(true, Ordering::Release);
+
+        if let Some(on_open) = self.on_open.as_ref() {
+            let on_open = on_open.clone();
+            self.handle.spawn(async move { on_open(()).await });
```

Comment:
```suggestion
            on_open(()).await;
```

Reply:
As far as I know, `await` runs the future to completion before continuing to the next `await` point or the end, that means the method wouldn't exit if the callback hangs or is very slow... The spawn method just spawns a new future in the executor and runs it to completion. The result is `()`, so we don't care about it anyways.

Reply:
Correct. Doesn't really matter either way. If we wanted to remove the handle parameter this change would be needed, otherwise it's fine as is.

Reply:
I don't know about the costs for a `Handle::current()` call. My initial thought was that if it was expensive it might be worth doing it just once and save the handle, but if inexpensive the alternative of just calling `Handle::current().spawn(...)` is more elegant.

Reply:
Either way

Reply:
Let's stick with the handle option for now ;)